javalib/src/main/scala/java/lang/Throwables.scala

```diff
@@ -49,49 +49,111 @@ private[lang] object StackTrace {
   }
 }
 
-class Throwable(s: String, private var e: Throwable)
+class Throwable protected (s: String,
+                           private[this] var e: Throwable,
+                           enableSuppression: scala.Boolean,
+                           writableStackTrace: scala.Boolean)
     extends Object
     with java.io.Serializable {
+
+  def this(message: String, cause: Throwable) =
+    this(message, cause, true, true)
+
   def this() = this(null, null)
+
   def this(s: String) = this(s, null)
+
   def this(e: Throwable) = this(if (e == null) null else e.toString, e)
 
-  private var stackTrace: Array[StackTraceElement] = _
+  private[this] var stackTrace: Array[StackTraceElement] = _
 
-  fillInStackTrace()
+  if (writableStackTrace)
+    fillInStackTrace()
 
-  def initCause(cause: Throwable): Throwable = {
-    e = cause
-    this
-  }
+  // We use an Array rather than, say, a List, so that Throwable does not
+  // depend on the Scala collections.
+  private[this] var suppressed: Array[Throwable] = _
 
-  def getMessage(): String = s
+  final def addSuppressed(exception: Throwable): Unit = {
 
-  def getCause(): Throwable = e
+    if (exception == null) {
+      // Yes, JVM message uses 'cannot' (sic) & has terminal period/full_stop!
+      throw new java.lang.NullPointerException(
+        "Cannot suppress a null exception.")
+    }
 
-  def getLocalizedMessage(): String = getMessage()
+    if (exception == this) {
+      throw new java.lang.IllegalArgumentException(
+        "Self-suppression not permitted")
+    }
+
+    if (enableSuppression) this.synchronized {
+      if (suppressed eq null) {
+        suppressed = Array(exception)
+      } else {
+        val length        = suppressed.length
+        val newSuppressed = new Array[Throwable](length + 1)
+        System.arraycopy(suppressed, 0, newSuppressed, 0, length)
+        newSuppressed(length) = exception
+        suppressed = newSuppressed
+      }
+    }
+  }
 
   def fillInStackTrace(): Throwable = {
-    this.stackTrace = StackTrace.currentStackTrace()
+    // currentStackTrace should be handling exclusion in its own
+    // critical section, but does not. So do
+    if (writableStackTrace) this.synchronized {
+      this.stackTrace = StackTrace.currentStackTrace()
+    }
     this
   }
 
+  def getCause(): Throwable = e
+
+  def getLocalizedMessage(): String = getMessage()
+
+  def getMessage(): String = s
+
   def getStackTrace(): Array[StackTraceElement] = {
     if (stackTrace eq null) {
-      stackTrace = Array.empty
-    }
-    stackTrace
+      Array.empty[StackTraceElement]
+    } else
+      this.synchronized {
+        stackTrace.clone
+      }
```

Comment:
This part in Scala.js does esentially the code for `fillInStackTrace` if you compare the code. Here if it is empty you just return empty rather than trying to fill it.

Reply:
This didn't work how I wanted it to. It was suppose to point to the `getStackTrace` method.

Reply:
After off-line discussion, I believe that Eric and I agreed to leave the code as it is.  
The revised code specifies the type of the empty array and follows the JVM
spec more closely by cloning stackTrace rather than returning the writable original.

@ekrich If you agree, could you take a moment and resolve this conversation.  If 
I have mis-characterized or misundertood our discussion, please advise.

Reply:
I think that sounds good if that is how you interpret the documents as well. If people have performance or other concerns the original is fine with me with a simple comment.

Reply:
Ah, a "simple comment" turned into a long tour through some tricky code!  

@ekrich If/when you have time, could you check the comment you recommended
and see if I described the situation so that another devo could follow it?  Thank you.

I can re-work it if it comes off as word salad....

Reply:
The strategy here is correct. Scala.js stores raw data in `fillInStackTrace()`, and only expands them into an `Array[StackTraceElement]` on first request. Scala Native eagerly creates the `StackTraceElement`s in `fillInStackTrace()`, so `getStackTrace()` doesn't have to do anything special.